### PR TITLE
auth0 JWT/JWKS authentication

### PR DESCRIPTION
**Pull request by goreRatzete**
I took the pr #188 and added the last missing bits.
* Tests
* Exception handling

Updated to 0.9 / latest commit. The old one was still using maven. 

I noticed that the other auth modules (auth, ldap) have their samples located in ktor-samples-auth. Should I move the jwt samples to the same location ?

### CODE REVIEW

File: ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+private suspend fun evaluate(token: String?, principal: Principal?, context: AuthenticationContext) {
+    val cause = when {
+        token == null -> NotAuthenticatedCause.NoCredentials
+        principal == null -> NotAuthenticatedCause.InvalidCredentials
+        else -> null
+    }
+    if (cause != null) {
+        context.challenge(JWTAuthKey, cause) {
+            call.respond(HttpStatusCode.Unauthorized)
```
Comment (cy6erGn0m) [inline ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: Shouldn't we provide 'WWW-Authenticate` header with `Bearer` challenge?

```
context.challenge(JWTAuthKey, NotAuthenticatedCause.InvalidCredentials) {
                it.success()
                call.respond(UnauthorizedResponse(HttpAuthHeader.bearerAuthChallenge(realm)))
            }

private fun HttpAuthHeader.Companion.bearerAuthChallenge(realm: String): HttpAuthHeader = 
    HttpAuthHeader.Parameterized("Bearer", mapOf(HttpAuthHeader.Parameters.Realm to realm))
```

You also can read https://www.scottbrady91.com/Kotlin/JSON-Web-Token-Verification-in-Ktor-using-Kotlin-and-Java-JWT

Also I am aware of parameters `error` and `error_description` that could be used to describe `cause` better so we can send only real if no credentials and error if the credentials are invalid.

File: ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import io.ktor.application.*
+import io.ktor.auth.*
+import io.ktor.http.*
+import io.ktor.response.*
+import java.security.interfaces.*
+import java.util.*
+
+private val JWTAuthKey: Any = "JWTAuth"
+
+data class JWTCredential(val payload: Payload) : Credential
```
Comment (cy6erGn0m) [inline ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: does `Payload` provide `hashCode` and `equals`?

File: ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    val payload = jwt.parsePayload()
+    val credentials = payload.let(::JWTCredential)
+    return credentials.let(validate)
+}
+
+private fun ApplicationCall.getAuthToken(): String? {
+    val headers = request.headers
+    val authHeader = headers[HttpHeaders.Authorization] ?: return null
+    val token = authHeader.removePrefix("Bearer ")
```
Comment (cy6erGn0m) [inline ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: See `parseAuthorizationHeader(headerValue: String)` and related

**Comment by cy6erGn0m** [comment]
I'd say we can keep all auth samples together

File: ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt (hunk 1 above)
Comment (goreRatzete) [reply ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: Regarding the `error` parameters: Do you mean as part of the Realm header?

File: ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt (hunk 2 above)
Comment (goreRatzete) [reply ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: No. PayloadImpl is a plain class, no `equals` or `hashCode.`

Comment (cy6erGn0m) [reply ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: `toString()`? why these classes are data classes?

Comment (goreRatzete) [reply ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: I removed the data class modifier in the latest commit.

**Comment by cy6erGn0m** [comment]
Ok, and finally the one one question remaining: is there something we can document? At last at two auth pipeline extension functions to help users understand what they should provide and why?

**Comment by goreRatzete** [comment]
Ok, should I make a page in ktorio.github.io ?

**Comment by cy6erGn0m** [comment]
Thanks! Yes, it would be great to have a page on ktorio.github.io!

**Comment by goreRatzete** [comment]
https://github.com/ktorio/ktorio.github.io/pull/21

**Pull request merged**